spinnaker-monitoring/prometheus_service.py

```diff
@@ -0,0 +1,209 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Implements metric service for interacting with Prometheus.
+
+Rather than pushing into prometheus, we'll let prometheus call us
+and collect on demand. However the base interface assumes it can call
+us so we'll stub that out with a no-op.
+
+To use this service, configure prometheus.yml as follows:
+
+scrape_configs:
+  - job_name: 'spinnaker'
+    static_configs:
+      - targets: ['localhost:8008']
+    metrics_path: '/prometheus_metrics'
+    honor_labels: true
+
+
+Where the localhost:8008 is the spinnaker-monitoring service.
+The 'honor_labels: true' uses the job and service labels injected from
+this service (which will be the spinnaker microservices the metrics came from)
+rather than the job and instance labels of this service which is
+what prometheus is scraping to collect the metrics.
+"""
+
+import collections
+import os
+
+import command_processor
+import spectator_client
+import util
+
+from prometheus_client import (
+    CONTENT_TYPE_LATEST,
+    generate_latest)
+
+from prometheus_client.core import (
+    GaugeMetricFamily,
+    CounterMetricFamily,
+    REGISTRY)
+
+
+InstanceRecord = collections.namedtuple(
+    'InstanceRecord', ['service', 'netloc', 'tags', 'data'])
+MetricInfo = collections.namedtuple('MetricInfo', ['kind', 'tags', 'records'])
+
+
+class PrometheusMetricsService(object):
+  """Implements monitoring service that implements a Prometheus client.
+
+  This service implements the Prometheus client library interface which
+  collects metrics in response to a collect() call.
+  """
+
+  def __init__(self, options):
+    self.__catalog = spectator_client.get_source_catalog(options['config_dir'])
+    options = util.merge_options_and_yaml_from_path(
+        options, os.path.join(options['config_dir'], 'prometheus.conf'))
+
+    self.__spectator = spectator_client.SpectatorClient(options)
+    self.__add_metalabels = config.get('prometheus_add_source_metalabels', True)
+    REGISTRY.register(self)
+
+  def __collect_instance_info(
+        self, service, name,
+        instance, metric_metadata, service_metadata, service_to_name_to_info):
+    """Creates a post payload for a DataDog time series data point.
+
+       See http://docs.datadoghq.com/api/?lang=python#metrics-post.
```

Comment:
good catch. originally copy pasted from datadog implementation to get the outline in place.